File: sdk/crewai/src/proceedgate_crewai/callback.py
"""
ProceedGate Callback for CrewAI
"""

from typing import Optional, Callable, Literal, Any
import re
from .client import ProceedGateClient
from .exceptions import FrictionRequiredError


class ProceedGateCallback:
    """
    CrewAI callback handler for ProceedGate cost governance.
    
    Tracks and controls costs during crew execution.
    """
    
    def __init__(
        self,
        api_key: str,
        workspace_id: str,
        base_url: str = "https://governor.proceedgate.dev",
        policy_id: str = "retry_friction_v1",
        on_friction: Literal["block", "warn"] = "warn",
    ):
        self.client = ProceedGateClient(api_key, workspace_id, base_url)
        self.policy_id = policy_id
        self.on_friction = on_friction
        
        self.total_cost = 0.0
        self.step_count = 0
        self.attempt_counts: dict[str, int] = {}
        
        # Event handlers
        self._on_gate_check: list[Callable] = []
        self._on_friction: list[Callable] = []
        self._on_cost_update: list[Callable] = []
    
    def on_gate_check(self, func: Callable) -> Callable:
        """Decorator to register gate check handler."""
        self._on_gate_check.append(func)
        return func
    
    def on_friction(self, func: Callable) -> Callable:
        """Decorator to register friction handler."""
        self._on_friction.append(func)
        return func
    
    def on_cost_update(self, func: Callable) -> Callable:
        """Decorator to register cost update handler."""
        self._on_cost_update.append(func)
        return func
    
    def _emit_gate_check(self, result: dict):
        for handler in self._on_gate_check:
            try:
                handler(result)
            except Exception:
                pass
    
    def _emit_friction(self, info: dict):
        for handler in self._on_friction:
            try:
                handler(info)
            except Exception:
                pass
    
    def _emit_cost_update(self, total: float):
        for handler in self._on_cost_update:
            try:
                handler(total)
            except Exception:
                pass
    
    def _get_attempt_count(self, key: str) -> int:
        count = self.attempt_counts.get(key, 0) + 1
        self.attempt_counts[key] = count
        return count
    
    def _parse_price(self, price_str: str) -> float:
        """Extract numeric value from price string like '0.004 USDC'."""
        match = re.search(r"([\d.]+)", price_str)
        if match:
            return float(match.group(1))
        return 0.0
    
    def gate_check(
        self,
        action: str,
        context: Optional[dict] = None,
    ) -> bool:
        """
        Check if an action is allowed.
        
        Returns True if allowed, raises or warns on friction.
        """
        attempt_key = f"{action}:{context.get('tool', 'default') if context else 'default'}"
        attempt = self._get_attempt_count(attempt_key)
        
        full_context = {
            "attempt_in_window": attempt,
            "window_seconds": 30,
            **(context or {}),
        }
        
        result = self.client.check(action, self.policy_id, full_context)
        
        self._emit_gate_check({
            "allowed": result.allowed,
            "decision_id": result.decision_id,
            "reason_code": result.reason_code,
            "price": result.policy.get("friction_price"),
        })
        
        if result.allowed:
            # Track cost
            price = self._parse_price(result.policy.get("friction_price", "0"))
            self.total_cost += price
            self.step_count += 1
            self._emit_cost_update(self.total_cost)
            return True
        
        # Friction required
        friction_info = {
            "decision_id": result.decision_id,
            "reason_code": result.reason_code,
            "price": result.policy.get("friction_price", "unknown"),
            "action": action,
        }
        self._emit_friction(friction_info)
        
        if self.on_friction == "block":
            raise FrictionRequiredError(
                f"Action '{action}' blocked. Friction required: {friction_info['price']}",
                decision_id=result.decision_id,
                reason_code=result.reason_code or "friction",
                price=friction_info["price"],
                action=action,
            )
        
        # Warn and continue
        print(f"[ProceedGate] Warning: Action '{action}' requires friction: {friction_info['price']}")
        return True
    
    # CrewAI callback interface
    def on_task_start(self, task: Any):
        """Called when a task starts."""
        self.gate_check("task_start", {"task": str(task)})
    
    def on_task_end(self, task: Any, output: Any):
        """Called when a task ends."""
        pass
    
    def on_tool_start(self, tool_name: str, tool_input: Any):
        """Called when a tool is invoked."""
        self.gate_check("tool_call", {"tool": tool_name})
    
    def on_tool_end(self, tool_name: str, tool_output: Any):
        """Called when a tool completes."""
        pass
    
    def on_llm_start(self, prompt: str):
        """Called before LLM invocation."""
        self.gate_check("llm_call", {
            "input_length": len(prompt),
            "confidence": 0.8,
        })
    
    def on_llm_end(self, response: str):
        """Called after LLM response."""
        pass
    
    def reset(self):
        """Reset tracking state."""
        self.total_cost = 0.0
        self.step_count = 0
        self.attempt_counts.clear()

File: sdk/langchain/src/callback.ts
/**
 * ProceedGate Callback Handler for LangChain
 * 
 * Intercepts LLM calls and tool invocations to enforce cost governance.
 */

import { BaseCallbackHandler } from '@langchain/core/callbacks/base';
import type { Serialized } from '@langchain/core/load/serializable';
import type { LLMResult } from '@langchain/core/outputs';
import type {
  ProceedGateCallbackHandlerOptions,
  CheckRequest,
  CheckResponse,
  FrictionInfo,
  GateCheckResult,
} from './types.js';

export class ProceedGateCallbackHandler extends BaseCallbackHandler {
  name = 'ProceedGateCallbackHandler';
  
  private apiKey: string;
  private workspaceId: string;
  private baseUrl: string;
  private policyId: string;
  private onFriction: 'block' | 'warn' | ((info: FrictionInfo) => Promise<boolean>);
  private trackTokens: boolean;

  private attemptCounts: Map<string, number> = new Map();
  private totalCost = 0;
  private eventListeners: Map<string, Array<(data: unknown) => void>> = new Map();

  constructor(options: ProceedGateCallbackHandlerOptions) {
    super();
    this.apiKey = options.apiKey;
    this.workspaceId = options.workspaceId;
    this.baseUrl = options.baseUrl ?? 'https://governor.proceedgate.dev';
    this.policyId = options.policyId ?? 'retry_friction_v1';
    this.onFriction = options.onFriction ?? 'warn';
    this.trackTokens = options.trackTokens ?? true;
  }

  /**
   * Register event listener
   */
  on(event: 'gate_check' | 'friction' | 'cost', callback: (data: unknown) => void): void {
    const listeners = this.eventListeners.get(event) ?? [];
    listeners.push(callback);
    this.eventListeners.set(event, listeners);
  }

  private emit(event: string, data: unknown): void {
    const listeners = this.eventListeners.get(event) ?? [];
    for (const listener of listeners) {
      try {
        listener(data);
      } catch {
        // Ignore listener errors
      }
    }
  }

  /**
   * Get total accumulated cost
   */
  getTotalCost(): number {
    return this.totalCost;
  }

  /**
   * Reset cost tracking
   */
  resetCost(): void {
    this.totalCost = 0;
    this.attemptCounts.clear();
  }

  private async gateCheck(action: string, context: Record<string, unknown>): Promise<GateCheckResult> {
    const attemptKey = `${action}:${context.tool ?? 'default'}`;
    const attempt = (this.attemptCounts.get(attemptKey) ?? 0) + 1;
    this.attemptCounts.set(attemptKey, attempt);

    const request: CheckRequest = {
      policy_id: this.policyId,
      action,
      actor: {
        id: `langchain-agent:${this.workspaceId}`,
        project: this.workspaceId,
      },
      context: {
        attempt_in_window: attempt,
        window_seconds: 30,
        ...context,
      },
    };

    const response = await fetch(`${this.baseUrl}/v1/governor/check`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${this.apiKey}`,
      },
      body: JSON.stringify(request),
    });

    const data = (await response.json()) as CheckResponse;

    const result: GateCheckResult = {
      allowed: data.allowed,
      decisionId: data.decision_id,
      proceedToken: data.allowed ? data.proceed_token : undefined,
      reasonCode: data.reason_code,
      frictionPrice: data.policy.friction_price,
    };

    this.emit('gate_check', result);
    return result;
  }

  private async handleFriction(info: FrictionInfo): Promise<boolean> {
    this.emit('friction', info);

    if (this.onFriction === 'block') {
      throw new Error(`ProceedGate: Action blocked. Friction required: ${info.price}`);
    }

    if (this.onFriction === 'warn') {
      console.warn(`[ProceedGate] Friction required for ${info.action}: ${info.price}`);
      return true; // Continue anyway
    }

    // Custom handler
    return this.onFriction(info);
  }

  /**
   * Called before LLM starts
   */
  async handleLLMStart(
    llm: Serialized,
    prompts: string[],
    _runId: string,
    _parentRunId?: string,
    extraParams?: Record<string, unknown>,
  ): Promise<void> {
    // Extract model name from serialized LLM or extraParams
    let model = 'unknown';
    if (extraParams?.invocation_params) {
      const params = extraParams.invocation_params as Record<string, unknown>;
      model = String(params.model_name ?? params.model ?? 'unknown');
    } else if ('kwargs' in llm && llm.kwargs) {
      const kwargs = llm.kwargs as Record<string, unknown>;
      model = String(kwargs.model_name ?? kwargs.model ?? 'unknown');
    }

    // Estimate input tokens (rough: 4 chars per token)
    const inputTokens = prompts.reduce((sum, p) => sum + Math.ceil(p.length / 4), 0);

    const result = await this.gateCheck('llm_call', {
      model: String(model),
      input_tokens: inputTokens,
      confidence: 0.8,
    });

    if (!result.allowed) {
      const shouldContinue = await this.handleFriction({
        decisionId: result.decisionId,
        reasonCode: result.reasonCode ?? 'friction',
        price: result.frictionPrice ?? 'unknown',
        action: 'llm_call',
      });

      if (!shouldContinue) {
        throw new Error(`ProceedGate: LLM call blocked`);
      }
    }
  }

  /**
   * Called after LLM ends
   */
  async handleLLMEnd(output: LLMResult): Promise<void> {
    if (this.trackTokens && output.llmOutput) {
      const usage = output.llmOutput.tokenUsage as {
        promptTokens?: number;
        completionTokens?: number;
      } | undefined;

      if (usage) {
        // Rough cost estimate (using GPT-4o pricing as baseline)
        const inputCost = ((usage.promptTokens ?? 0) / 1000) * 0.0025;
        const outputCost = ((usage.completionTokens ?? 0) / 1000) * 0.01;
        this.totalCost += inputCost + outputCost;
        this.emit('cost', this.totalCost);
      }
    }
  }

  /**
   * Called before tool starts
   */
  async handleToolStart(
    tool: Serialized,
    input: string,
    _runId: string,
  ): Promise<void> {
    const toolName = tool.id?.[tool.id.length - 1] ?? 'unknown';

    const result = await this.gateCheck('tool_call', {
      tool: toolName,
      confidence: 0.9,
    });

    if (!result.allowed) {
      const shouldContinue = await this.handleFriction({
        decisionId: result.decisionId,
        reasonCode: result.reasonCode ?? 'friction',
        price: result.frictionPrice ?? 'unknown',
        action: 'tool_call',
        tool: toolName,
      });

      if (!shouldContinue) {
        throw new Error(`ProceedGate: Tool ${toolName} blocked`);
      }
    }
  }
}
